### PR TITLE
front: intervals editor, fix bugs

### LINKED ISSUE #4557: intervals editor: fix bugs
Globally fix the intervals editor. Here is an uncomplete list of bugs :
- [x] The values of the intervals must be displayed if there's enough space to do so
- [x] Ticks must be responsive and adapt dynamically to the size of the screen (the number of ticks must change).
- [x] Numbers of digits to be settable.
- [x] Non mandatory max interval value sticking to distance if distance = max.
- [x] Debounce on input change for metadata.
- [x] An empty interval can be selected
- [x] Vertical scale should not be displayed if the value of the intervals can have be in several units or is a string
- [x] Default value should be compulsory
- [ ] Documentation

### PR DESCRIPTION
**Pull request by clarani**
closes #4557 

I recommend to review this PR commit by commit :)

### CODE REVIEW

File: front/src/common/BootstrapSNCF/FormSNCF/DebouncedNumberInputSNCF.tsx [hunk 1]
```diff
@@ -0,0 +1,68 @@
+import React, { useEffect, useState } from 'react';
+import InputSNCF from 'common/BootstrapSNCF/InputSNCF';
+
+type DebouncedNumberInputSNCFProps = {
+  input: number;
+  label: string;
+  setInput: (number: number) => void;
```
Comment (SharglutDev) [inline front/src/common/BootstrapSNCF/FormSNCF/DebouncedNumberInputSNCF.tsx]: Maybe better with `(input: number)` or `(inputValue: number)` ? :)

**Review by Uriel-Sautron — APPROVED** [review]

File: front/src/common/IntervalsEditor/IntervalsEditor.tsx
Comment (Uriel-Sautron) [inline front/src/common/IntervalsEditor/IntervalsEditor.tsx:1]: You can change this:
export const IntervalsEditor: React.FC<IntervalsEditorProps> = (props)
to:
export const IntervalsEditor = (props: IntervalsEditorProps)

**Review by Uriel-Sautron — COMMENTED** [review]
Very good job, just a comment

File: front/src/common/IntervalsEditor/IntervalsEditorCommonForm.tsx [hunk 2]
```diff
@@ -19,51 +20,68 @@ const IntervalsEditorCommonForm = ({
...
-    setData(fixedResults);
+  const [begin, setBegin] = useState<number>(interval.begin);
+  const [end, setEnd] = useState(interval.end);
+
+  useEffect(() => {
+    setBegin(interval.begin);
+    setEnd(interval.end);
+  }, [interval]);
+
+  /** Resize segment from inputs */
```
Comment (SharglutDev) [inline front/src/common/IntervalsEditor/IntervalsEditorCommonForm.tsx]: Is this comment needed ? The function name is pretty clear I think

**Review by SharglutDev — COMMENTED** [review]
Great work :) Just a couple of suggestions

File: front/src/common/IntervalsEditor/IntervalsEditorCommonForm.tsx (hunk 2 above)
Comment (clarani) [reply front/src/common/IntervalsEditor/IntervalsEditorCommonForm.tsx]: Done !

File: front/src/common/BootstrapSNCF/FormSNCF/DebouncedNumberInputSNCF.tsx (hunk 1 above)
Comment (clarani) [reply front/src/common/BootstrapSNCF/FormSNCF/DebouncedNumberInputSNCF.tsx]: Done !

File: front/src/common/IntervalsEditor/IntervalsEditor.tsx
Comment (clarani) [reply front/src/common/IntervalsEditor/IntervalsEditor.tsx:1]: Done, thanks for this comment !

File: front/src/common/IntervalsEditor/IntervalsEditor.tsx [hunk 3]
```diff
@@ -196,11 +197,16 @@ export const IntervalsEditor: React.FC<IntervalsEditorProps> = (props) => {
             onResize={(index, gap, finalized) => {
               setMode(!finalized ? 'resizing' : null);
               try {
-                const { result, newIndexMapping } = resizeSegment(data, index, gap, 'end');
+                const { result, newIndexMapping } = resizeSegment(data, index, gap, 'end', false);
+                const fixedResults = fixLinearMetadataItems(result?.filter(notEmpty), totalLength, {
+                  fieldName: 'value',
+                  defaultValue,
+                });
+
```
Comment (SharglutDev) [inline front/src/common/IntervalsEditor/IntervalsEditor.tsx]: Not sure if it is wanted but when you resize the end of the "existing segment" left to the last segment (the empty one) of the editor, this last one is transformed into an "existing segment". Seems like a strange behaviour.
It means that if we want to only resize the before last segment, it transforms the last one and we have to delete it afterwards.

Comment (SharglutDev) [reply front/src/common/IntervalsEditor/IntervalsEditor.tsx]: Ok nevermind, it seems that it's fixed in the next commit :)

File: front/src/common/IntervalsEditor/utils.ts [hunk 4]
```diff
@@ -40,3 +39,57 @@ export function createEmptySegmentAt<T>(
...
+  }
+
+  return linearMetadata
+    .map((item, index) => {
+      if (index === indexToRemove - 1 && shouldMergeIntoPreviousSegment) {
+        if (shouldMergeIntoTheNextSegment) {
+          // both adjacent segments are empty
+          return [{ ...item, end: linearMetadata[indexToRemove + 1].end }];
+        }
+        // merge the segment to remove into th previous segment
```
Comment (SharglutDev) [inline front/src/common/IntervalsEditor/utils.ts]: typo here :)

File: front/src/common/IntervalsEditor/utils.ts [hunk 5]
```diff
@@ -40,3 +39,57 @@ export function createEmptySegmentAt<T>(
...
+        return [{ ...item, end: linearMetadata[indexToRemove].end }];
+      }
+      if (index === indexToRemove) {
+        if (shouldMergeIntoPreviousSegment || shouldMergeIntoTheNextSegment) return [];
+        // remove the selected segment without merging it into adjacent ones
+        return [{ ...item, value: emptyValue, ...(defaultUnit ? { unit: defaultUnit } : {}) }];
+      }
+      if (index === indexToRemove + 1 && shouldMergeIntoTheNextSegment) {
+        if (shouldMergeIntoPreviousSegment) return [];
+        // merge  the segment to remove into next segment
```
Comment (SharglutDev) [inline front/src/common/IntervalsEditor/utils.ts]: one extra space after "merge"

**Review by SharglutDev — APPROVED** [review]
Great work :)

File: front/src/common/IntervalsEditor/utils.ts (hunk 4 above)
Comment (clarani) [reply front/src/common/IntervalsEditor/utils.ts]: Done ! Thanks

File: front/src/common/IntervalsEditor/utils.ts (hunk 5 above)
Comment (clarani) [reply front/src/common/IntervalsEditor/utils.ts]: Done !

File: front/src/common/BootstrapSNCF/FormSNCF/DebouncedNumberInputSNCF.tsx [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+  min = 0,
+}: DebouncedNumberInputSNCFProps) => {
+  const [value, setValue] = useState<number | null>(input);
+  const [debouncedValue, setDebouncedValue] = useState<number | null>(null);
+
+  useEffect(() => {
+    setValue(input);
+    setDebouncedValue(null);
+  }, [input]);
+
```
Comment (alexandredamiron) [inline front/src/common/BootstrapSNCF/FormSNCF/DebouncedNumberInputSNCF.tsx:29]: You can use the useDebounce or useDebouncedFunc present in utils/helpers

File: front/src/common/IntervalsDataViz/Scales.tsx [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+  begin: number;
+  end: number;
+  className?: string;
+}) => {
+  const [ticksCount, setTicksCount] = useState<number>(10);
+
+  const inf = roundNumber(begin, true);
+  const sup = roundNumber(end, false);
+  const step = roundNumber((end - begin) / ticksCount / 2, true);
+
```
Comment (alexandredamiron) [inline front/src/common/IntervalsDataViz/Scales.tsx:20]: Same, you can use useDebounce or useDebouncedFunc in utils/helpers

File: front/src/common/IntervalsDataViz/IntervalItem.tsx [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+  min: number;
+  max: number;
+  resizing: { index: number | null; startAt: number } | null;
+  segment: LinearMetadataItem & { index: number };
+  setDraginStartAt: (dragingStartAt: number) => void;
+  setHoverAtx: (hoverAtx: number) => void;
+  setResizing: (resizing: { index: number | null; startAt: number } | null) => void;
+  wrapper: React.MutableRefObject<HTMLDivElement | null>;
+}
+
```
Comment (alexandredamiron) [inline front/src/common/IntervalsDataViz/IntervalItem.tsx:21]: no need to disable next line @typescript-eslint/no-explicit-any, y<T extends { [key: string]: string | number }>

Comment (clarani) [reply front/src/common/IntervalsDataViz/IntervalItem.tsx:21]: Done !

File: front/src/common/BootstrapSNCF/FormSNCF/DebouncedNumberInputSNCF.tsx (hunk 6 above)
Comment (clarani) [reply front/src/common/BootstrapSNCF/FormSNCF/DebouncedNumberInputSNCF.tsx:29]: Done, thanks !! ;)

File: front/src/common/IntervalsDataViz/Scales.tsx (hunk 7 above)
Comment (clarani) [reply front/src/common/IntervalsDataViz/Scales.tsx:20]: I think I cannot do this here, as I use window.resize event, what do you think ?

**Review by alexandredamiron — APPROVED** [review]

File: front/src/common/IntervalsDataViz/Scales.tsx (hunk 7 above)
Comment (alexandredamiron) [reply front/src/common/IntervalsDataViz/Scales.tsx:20]: indeed, helpers are not fitted. A general function for resize on scale should be useful

**Pull request merged**